pkg/generate/code/equal_test.go

```diff
@@ -0,0 +1,173 @@
+// License is located at
+//
+//	 http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package code_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/aws-controllers-k8s/code-generator/pkg/generate/code"
+	"github.com/aws-controllers-k8s/code-generator/pkg/testutil"
+)
+
+func TestCompareResource_ApigatewayV2_RouteSettings(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	g := testutil.NewGeneratorForService(t, "apigatewayv2")
+
+	tdef := testutil.GetTypeDefByName(t, g, "RouteSettings")
+	require.NotNil(tdef)
+
+	expected := `	if ackcompare.HasNilDifference(a.DataTraceEnabled, b.DataTraceEnabled) {
+		return false
+	} else if a.DataTraceEnabled != nil && b.DataTraceEnabled != nil {
+		if *a.DataTraceEnabled != *b.DataTraceEnabled {
+			return false
+		}
+	}
+	if ackcompare.HasNilDifference(a.DetailedMetricsEnabled, b.DetailedMetricsEnabled) {
+		return false
+	} else if a.DetailedMetricsEnabled != nil && b.DetailedMetricsEnabled != nil {
+		if *a.DetailedMetricsEnabled != *b.DetailedMetricsEnabled {
+			return false
+		}
+	}
+	if ackcompare.HasNilDifference(a.LoggingLevel, b.LoggingLevel) {
+		return false
+	} else if a.LoggingLevel != nil && b.LoggingLevel != nil {
+		if *a.LoggingLevel != *b.LoggingLevel {
+			return false
+		}
+	}
+	if ackcompare.HasNilDifference(a.ThrottlingBurstLimit, b.ThrottlingBurstLimit) {
+		return false
+	} else if a.ThrottlingBurstLimit != nil && b.ThrottlingBurstLimit != nil {
+		if *a.ThrottlingBurstLimit != *b.ThrottlingBurstLimit {
+			return false
+		}
+	}
+	if ackcompare.HasNilDifference(a.ThrottlingRateLimit, b.ThrottlingRateLimit) {
+		return false
+	} else if a.ThrottlingRateLimit != nil && b.ThrottlingRateLimit != nil {
+		if *a.ThrottlingRateLimit != *b.ThrottlingRateLimit {
+			return false
+		}
+	}
+`
+	assert.Equal(
+		expected,
+		code.CheckEqualType(
+			tdef, "a", "b", 1,
+		),
+	)
+}
+
+func TestCompareResource_ECR_Repository_SDK(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	g := testutil.NewGeneratorForService(t, "ecr")
+
+	tdef := testutil.GetTypeDefByName(t, g, "Repository")
+	require.NotNil(tdef)
+
+	expected := `	if ackcompare.HasNilDifference(a.CreatedAt, b.CreatedAt) {
+		return false
+	} else if a.CreatedAt != nil && b.CreatedAt != nil {
+		if *a.CreatedAt != *b.CreatedAt {
+			return false
+		}
+	}
+	if ackcompare.HasNilDifference(a.ImageScanningConfiguration, b.ImageScanningConfiguration) {
+		return false
+	} else if a.ImageScanningConfiguration != nil && b.ImageScanningConfiguration != nil {
+		if !ImageScanningConfiguration(a.ImageScanningConfiguration, b.ImageScanningConfiguration) {
```

Comment:
"!ImageScanningConfiguration(a.ImageScanningConfiguration, b.ImageScanningConfiguration)"

this looks incorrect. Is there a function with same name as shape?

Reply:
>Is there a function with same name as shape?

Yes, for each type we generate a function (with same name as the type) that will be used to check the equality of two instances. The reason why we don't add any prefix to the function names, is that the package it self is called `equal`, so when importing this package in `delta.go`, we will be using it with something like `equal.ImageScanningConfiguration(a, b)`